contracts/src/zaps/UniV3Zap.sol

```diff
@@ -0,0 +1,937 @@
+    receive() external payable {
+        // Ensures that the zap is receiving ether inside the context of another
+        // call. This means that the ether transfer should be the result of a
+        // Hyperdrive trade or unwrapping WETH.
+        if (!_reentrancyGuardEntered()) {
+            revert InvalidTransfer();
+        }
+    }
+
+    /// LPs ///
+
+    /// @notice Executes a swap on Uniswap and uses the proceeds to add
+    ///         liquidity on Hyperdrive.
+    /// @param _hyperdrive The Hyperdrive pool to open the long on.
+    /// @param _minLpSharePrice The minimum LP share price the LP is willing
+    ///        to accept for their shares. LPs incur negative slippage when
+    ///        adding liquidity if there is a net curve position in the market,
+    ///        so this allows LPs to protect themselves from high levels of
+    ///        slippage. The units of this quantity are either base or vault
+    ///        shares, depending on the value of `_options.asBase`.
+    /// @param _minApr The minimum APR at which the LP is willing to supply.
+    /// @param _maxApr The maximum APR at which the LP is willing to supply.
+    /// @param _hyperdriveOptions The options that configure how the Hyperdrive
+    ///        operation is settled.
+    /// @param _zapInOptions The options that configure how the zap will be
+    ///        settled.
+    /// @return lpShares The LP shares received by the LP.
+    function addLiquidityZap(
+        IHyperdrive _hyperdrive,
+        uint256 _minLpSharePrice,
+        uint256 _minApr,
+        uint256 _maxApr,
+        IHyperdrive.Options calldata _hyperdriveOptions,
+        IUniV3Zap.ZapInOptions calldata _zapInOptions
+    ) external payable nonReentrant returns (uint256 lpShares) {
+        // Validate the zap parameters.
+        bool shouldConvertToETH = _validateZapIn(
+            _hyperdrive,
+            _hyperdriveOptions,
+            _zapInOptions
+        );
+
+        // Zap the funds that will be used to add liquidity and approve the pool
+        // to spend these funds.
+        uint256 proceeds = _zapIn(_zapInOptions, shouldConvertToETH);
+
+        // If the deposit isn't in ETH, we need to set an approval on Hyperdrive.
+        if (!shouldConvertToETH) {
+            // NOTE: We increase the required approval amount by 1 wei so that the
+            // pool ends with an approval of 1 wei. This makes future approvals
+            // cheaper by keeping the storage slot warm.
+            ERC20(_zapInOptions.swapParams.path.tokenOut()).forceApprove(
+                address(_hyperdrive),
+                proceeds + 1
+            );
+        }
+
+        // Add liquidity using the proceeds of the trade. If the vault shares
+        // token is a rebasing token, the proceeds amount needs to be converted
+        // to vault shares.
+        if (!_hyperdriveOptions.asBase && _zapInOptions.isRebasing) {
```

Comment:
okay so even if the base token is rebasing there is no need to convert to shares bc it is base. sorry, this confused me for a second so im just typing it out.

Reply:
Yeah that's right. And then the other two cases are:

1. Non-rebasing and `asBase = true`: The proceeds amount is in base, so no conversion is necessary.
2. Non-rebasing and `asBase = false`: The proceeds amount is in vault shares, so no conversion is necessary.